### DISCUSSION TITLE
Typescript error with form and handleSubmit

### DISCUSSION THREAD

**Discussion by SalahAdDin** [react-hook-form/react-hook-form#8020] [Q&A]
Hi guys, 

I'm using this for handling my form states, together with **Typescript**, but now, I'm getting an error with type checking:
`Promise-returning function provided to attribute where a void return was expected`.

My code is this:
```
const { dispatch } = useContext(Context);

  const form = useForm<ICreateTaskForm>({
    defaultValues: { title: '', description: '' },
    resolver: yupResolver(schema),
  });
  const { control, handleSubmit, reset } = form;

  const onSubmit = handleSubmit(({ title, description }: ICreateTaskForm) => {
    dispatch({
      type: ActionTypes.createTask,
      payload: { id: nanoid(), title, description, status: 'ToDo' },
    });
    reset();
  });

return(<form onSubmit={onSubmit}>
        <TextInput name="title" control={control} label="Title" required />
        <TextInput
          name="description"
          control={control}
          label="Description"
          required
          multiline
          rows={5}
        />
        <Button type="submit" variant="contained" startIcon={<Add />} fullWidth>
          Add
        </Button>
      </form>)
      
```

I know `handleSubmit` returns a  `Promise`, but I don't know how to solve this problem.

Can anyone help me?

Thanks

**Comment by sergey-sivchenko** [comment]
Same issue for me coming from [@typescript-eslint/eslint-plugin](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/no-misused-promises.md).

Another related case would be not spreading the registered field properties:

```javascript
const field = register("field");
...
<input
  ref={field.ref}
  name={field.name}
  onBlur={field.onBlur} // Same error here as onBlur returns Promise<void | boolean>
  onChange={field.onChange} // Same error here as onChange returns Promise<void | boolean>
  type="text"
/>
```

**Comment by Moshyfawn** [comment]
Hey!

Please, take a look at [this CSB](https://codesandbox.io/s/sleepy-snyder-yipnc4?file=/src/App.tsx) and possibly tweak to match your use-case, where you get the error.

**Reply by sergey-sivchenko** [reply]
Hey!
Sorry for not providing the code example, but as far as I understand [CodeSandbox does not support custom eslint config](https://codesandbox.io/docs/environment). So, you will not be able to see the errors mentioned above in CSB. However, I created the [example based on your CSB](https://codesandbox.io/s/cocky-hill-vez6d2?file=/src/App.tsx), which produce same errors locally:

<img width="997" alt="Screenshot 2022-03-25 at 11 46 14" src="https://user-images.githubusercontent.com/50201905/160086815-c87e1fc3-7b35-445e-b2b0-4166ba59801b.png">
<img width="986" alt="Screenshot 2022-03-25 at 11 46 59" src="https://user-images.githubusercontent.com/50201905/160086838-bf068737-5a02-421b-b635-3ac2c69cb9a2.png">

**Summary of changes**

1. Changed `tsconfig.json` to the [default from CRA template typescript](https://create-react-app.dev/docs/getting-started#creating-a-typescript-app)
2. [Extended default CRA ESLint config](https://create-react-app.dev/docs/setting-up-your-editor#extending-or-replacing-the-default-eslint-config) with [Typescript ESLint Linting with Type Information](https://typescript-eslint.io/docs/linting/type-linting)
3. Replaced spread from `title` field by direct props passing

**Reply by Moshyfawn** [reply]
Thx for the example! Will take a look

**Reply by sergey-sivchenko** [reply]
Here is the situation as I understand it according to the types:

- `@typescript-eslint` can [provide lint rules with access to all of the type information that TypeScript knows about your codebase](https://typescript-eslint.io/docs/linting/type-linting), which is fine
- its `no-misused-promises` [rule forbids providing Promises to logical locations in places where the TypeScript compiler allows them but they are not handled properly](https://typescript-eslint.io/rules/no-misused-promises), which is fine
- `react`'s `form` accepts `onSubmit?: FormEventHandler<T> | undefined;`, which is essentially `EventHandler` and it is not async function, which is fine
- `react`'s `input` accepts `onChange?: ChangeEventHandler<T> | undefined;`, which is essentially `EventHandler` and it is not async function, which is fine
- `react-hook-form`'s `handleSubmit` creates a submit function, which is async function and returns `Promise<void>`, which is fine by itself
- `react-hook-form`'s `register` creates `ChangeHandler`, which is async function and returns `Promise<void | boolean>`, which is fine by itself

Despite the fact everything might work as expected from the code point of view, combining together `react`'s `form` and `react-hook-form`'s submit function (as well as `react`'s `input` and `react-hook-form`'s `ChangeHandler`) we are getting a combination of regular and async functions, which leads to the situation when we have not properly handled Promises. As a result we have an error from `@typescript-eslint`.

Please, feel free to correct me if I'm wrong somewhere as I am neither an expert in Typescript nor in `react-hook-form`.

**Comment by egerpe-ats** [comment]
The documentation of [@typescript-eslint](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/no-misused-promises.md#checksvoidreturn) suggests disable the rule only for that particular case. Maybe it's not a solution but it does the trick for me.

**Reply by SalahAdDin** [reply]
Of course we are using this trick, we report this here cause it should be a better solution.

**Comment by leapful** [comment]
@SalahAdDin , @sergey-sivchenko :

Since it's just a linting error from eslint rule, not typescript error, the last thing you can do is following its best practice to avoid potential issue, may be a promise handling wrapper like this:

```
function onPromise<T>(promise: (event: SyntheticEvent) => Promise<T>) {
  return (event: SyntheticEvent) => {
    if (promise) {
      promise(event).catch((error) => {
        console.log("Unexpected error", error);
      });
    }
  };
}
```

Demo: https://codesandbox.io/s/serene-grass-vehq90?file=/src/App.tsx

**Reply by SalahAdDin** [reply]
> @SalahAdDin , @sergey-sivchenko :
> 
> Since it's just a linting error from eslint rule, not typescript error, the last thing you can do is following its best practice to avoid potential issue, may be a promise handling wrapper like this:
> 
> ```
> function onPromise<T>(promise: (event: SyntheticEvent) => Promise<T>) {
>   return (event: SyntheticEvent) => {
>     if (promise) {
>       promise(event).catch((error) => {
>         console.log("Unexpected error", error);
>       });
>     }
>   };
> }
> ```
> 
> Demo: https://codesandbox.io/s/serene-grass-vehq90?file=/src/App.tsx

Thank you!, that notation is quite interesting, could you explain it please? Why are not you using arrow functions?

**Reply by leapful** [reply]
@SalahAdDin : Sorry for the late reply. The reason of using "function" instead of "arrow function" is just for comforting Typescript checking as with "arrow function", Typescript is not sure that "<" will open a JSX tag or generic definition.

You still can achieve generic type with "arrow function" but with some weird syntax like:

```
const fn = <T, >() => {};

const fn = <T extends unknown>() => {}
```

but I would not recommend this way as it's confusing.

Referrer: https://stackoverflow.com/questions/41112313/how-to-use-generics-with-arrow-functions-in-typescript-jsx-with-react

**Reply by SalahAdDin** [reply]
> @SalahAdDin : Sorry for the late reply. The reason of using "function" instead of "arrow function" is just for comforting Typescript checking as with "arrow function", Typescript is not sure that "<" will open a JSX tag or generic definition.
> 
> You still can achieve generic type with "arrow function" but with some weird syntax like:
> 
> ```
> const fn = <T, >() => {};
> 
> const fn = <T extends unknown>() => {}
> ```
> 
> but I would not recommend this way as it's confusing.
> 
> Referrer: https://stackoverflow.com/questions/41112313/how-to-use-generics-with-arrow-functions-in-typescript-jsx-with-react

Actually, I tried to translate it to an arrow function, I couldn't.

**Reply by trungh13** [reply]
> @SalahAdDin , @sergey-sivchenko :
> 
> Since it's just a linting error from eslint rule, not typescript error, the last thing you can do is following its best practice to avoid potential issue, may be a promise handling wrapper like this:
> 
> ```
> function onPromise<T>(promise: (event: SyntheticEvent) => Promise<T>) {
>   return (event: SyntheticEvent) => {
>     if (promise) {
>       promise(event).catch((error) => {
>         console.log("Unexpected error", error);
>       });
>     }
>   };
> }
> ```
> 
> Demo: https://codesandbox.io/s/serene-grass-vehq90?file=/src/App.tsx

Amazing workaround.

For some want to have help neater code then you can make it as helper then 
```
const { handleSubmit } = useForm()
const onSubmitHandler = () =>{
  // do something
}
....
<form onSubmit={onPromise(handleSubmit(onSubmitHandler))}>
```

**Reply by jamesarosen** [reply]
I love this solution!

I'll offer a few minor edits for clarity:

```ts
function withExceptionCapturing<T>(handler: (event: SyntheticEvent) => Promise<T>) {
  return (event: SyntheticEvent) => {
    if (handler) {
      handler(event).catch((error) => {
        console.log("Unexpected error", error);
      });
    }
  };
}
```

First, `onPromise` is quite vague. I changed it to `withExceptionCapturing` to describe the idea that we're taking in a handler and returning a handler _with exception capturing_.

Second, the argument to `withExceptionCapturing` isn't a `Promise`; it's a `SyntheticEvent => Promise` (a function that takes a `SyntheticEvent` and returns a `Promise`). Thus, I changed the name to `handler`.

You could go further and say that this doesn't even have anything to do with event handling or react-hook-forms at all. It's just about capturing dangling promises from functions. That might look like this:

```ts
function withExceptionCapturing<S, T extends any[]>(fn: (...rest: T) => Promise<S>) {
  return (...args: T) => {
    fn(...args).catch((error) => {
      console.log('Unexpected error', error);
    });
  };
}
```

And you'd use it the same way:

```tsx
<form onSubmit={withExceptionCapturing(handleSubmit(onSubmitHandler))}>
```

But you could also use that one in other contexts:

```ts
// (number) => Promise<string>
function aSmallNumber(n: number): Promise<string> {
  return new Promise((resolve, reject) => {
    if (n < 0) reject("Can't be negative");
    if (n < 10) resolve('OK');
    reject('Too big!');
  });
}

// also (number) => Promise<string>
const aSmallNumberWithExceptionCapturing = withExceptionCapturing(aSmallNumber);

aSmallNumberWithExceptionCapturing(7);
```

**Comment by nirtamir2** [comment]
I notice this issue too.
Can we somehow use another exposed function that does not returns a promise / change https://github.com/react-hook-form/react-hook-form/blob/8028e4d9421ce42acd8079f9c51b8fcfdac093ce/src/types/form.ts#L597 type to not return a promise?
It seems that all the solutions so far are just ignoring this ESLint rule / have non-idiomatic code

**Comment by paleite** [comment]
I suspect this is a bug in react-hook-form.
Here's the workaround:

Before:

```
      <form onSubmit={handleSubmit(onSubmit)}>
```

After:

```
      <form onSubmit={(...args) => void handleSubmit(onSubmit)(...args)}>
```

The bug: `onSubmit` expects a sync function that returns void, but `handleSubmit` is an async function returning void, hence why ESLint (correctly) identifies this as a misused Promise, and that's also why fixing it so ESLint doesn't complain (e.g. `<form onSubmit={void handleSubmit(onSubmit)}>`) breaks `handleSubmit` in a way that `preventDefault` is never called.

**Reply by sergey-sivchenko** [reply]
I would not say it is a bug. It is more misalignment in types between `react-hook-form` and `react` types (I suppose). Passing async function to `onSubmit` is not bad. The typing just warn that the result will not be handled properly. So, it is a problem of typing - yes, bug in logic - I believe no.

**Reply by paleite** [reply]
The reason I say there's a bug in the logic is because `<form onSubmit={void handleSubmit(onSubmit)}>` doesn't prevent the default-event, causing the form to be sent like a regular HTML-form, as well as sometimes calling the onSubmit-function (if the server is slow enough).
This wouldn't happen in a sync function.

**Reply by leapful** [reply]
@paleite : No, the code ```<form onSubmit={void handleSubmit(onSubmit)}>``` is not correct that why it does not prevent default from the event.

If you look at the implementation of handleSubmit, you will see that it's an high-order function, which means it takes 2 callback functions "onValid" and "onInvalid" as arguments and return another callback function for ```<form onSubmit>```.

https://github.com/react-hook-form/react-hook-form/blob/e7e449d18229c84db456df8e8223cd7a821de315/src/logic/createFormControl.ts#L1017-L1025

Calling "void handleSubmit(onSubmit)" only executes the function "handleSubmit" immediately and returns "undefined".

Referrer: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/void

**Reply by paleite** [reply]
Yes, that sounds about like what I had imagined. What should happen: call a sync function which first calls `preventDefault` and then executes the callbacks asynchronously. What it does instead: calls an async function (incorrect), which won't be resolved until the event-loop has called it, but by the time that happens, the form has already been submitted "via HTML".

What should happen instead is to call `preventDefault` first, then execute the async code and then return void.

I've posted an example below. Please note how `preventDefault` is executed in the sync function, and the async code is executed, awaited, and not returned. Resulting in `handleSubmit` correctly returning void.

```typescript
const handleSubmit: UseFormHandleSubmit<TFieldValues> =
  (onValid, onInvalid) =>
  // Sync, returns void
  (e): void => {
    if (e) {
      e.preventDefault && e.preventDefault();
      e.persist && e.persist();
    }

    // Async IIFE. Executes async functions `onValid` and `onInvalid`
    (async (event) => {
      let hasNoPromiseError = true;
      let fieldValues: any = cloneObject(_formValues);
      /** remaining code containing `await`s */
    })(e);
  };
```

**Reply by paleite** [reply]
https://github.com/react-hook-form/react-hook-form/commit/7bfa3747f42ee648d0440c98d37be832a98805f2 <- this is what `handleSubmit`'s return type should actually be

**Reply by leapful** [reply]
@paleite : The problem is not related to "handleSubmit" or the callback of "onSubmit" from ```<form onSubmit={callback}>``` of react-hook-form, it's from your usage of calling "void" syntax.


In the previous comment you claimed that calling:

``` <form onSubmit={(...args) => void handleSubmit(onSubmit)(...args)}>``` will help the "onSubmit" performs correctly "preventDefault" but ```<form onSubmit={void handleSubmit(onSubmit)}>``` will not perform "preventDefault"

The reason is simple: first sample is still a callback function is bind to "onSubmit" of ```<form>``` and every time form is submitting it will be called with prevent default from the callback result of "handleSubmit" while the second sample (calling "void") already executed the callback function and return "undefined" to "onSubmit" of ```<form>```


There is nothing wrong with using async callback for "onSubmit":

```
const onSubmit = async (event) => {
    if (event) {
      event.preventDefault();
      console.log("Prevent default");
    }
  };

<form onSubmit={onSubmit}>

 // or high order function that return a
const makeOnSubmit = (onValid, onInvalid) => async (event) => {
    if (event) {
      event.preventDefault();
      console.log("Prevent default");
    }
  
    onValid();
}

<form onSubmit={makeOnSubmit(someOnValidCallback)}>
```


Here is a working demo with using directly async callback for "onSubmit": https://codesandbox.io/s/sleepy-hopper-spj5ii?file=/src/App.tsx

**Reply by SalahAdDin** [reply]
So, what's the best practice here?

**Reply by cleisonmp** [reply]
> [7bfa374](https://github.com/react-hook-form/react-hook-form/commit/7bfa3747f42ee648d0440c98d37be832a98805f2) <- this is what `handleSubmit`'s return type should actually be

can you open a PR for this? worked like a charm for me, though not sure if it would break something else.

**Reply by kdubb** [reply]
> There is nothing wrong with using async callback for "onSubmit":

If I'm not mistaken, `handleSubmit` is a higher order function returning a function specifically for use in `onSubmit`. According to the docs & type signature `onSubmit` does not handle `async` or `Promise` returning functions.

Just for continuity's sake, `handleSubmit` should be returning functions that match the `onSubmit` signature but the larger issue is that anybody using `onSubmit={handleSubmit(mySubmitHandle)}` is leaking a Promise (i.e. errors produced by the promise are not handled). This is exactly why eslint  is producing a warning.

So @paleite's suggestion of "fixing the signature" seems to be correct as long as `handleSubmit` always returns a function that returns `void` (as it seemingly should).

**Reply by SalahAdDin** [reply]
> > There is nothing wrong with using async callback for "onSubmit":
> 
> If I'm not mistaken, `handleSubmit` is a higher order function returning a function specifically for use in `onSubmit`. According to the docs & type signature `onSubmit` does not handle `async` or `Promise` returning functions.
> 
> Just for continuity's sake, `handleSubmit` should be returning functions that match the `onSubmit` signature but the larger issue is that anybody using `onSubmit={handleSubmit(mySubmitHandle)}` is leaking a Promise (i.e. errors produced by the promise are not handled). This is exactly why **eslint** is producing a warning.
> 
> So @paleite's suggestion of "fixing the signature" seems to be correct as long as `handleSubmit` always returns a function that returns `void` (as it seemingly should).

Good to know, thank you.

**Reply by leapful** [reply]
@kdubb 

Rule `no-misused-promises` with option `checksVoidReturn` (option for the case in this discussion) is to prevent a asynchronous callback to be passed as an argument where the logic for that argument want call a synchronous callback, something like:

```
const setSafeMode = async () => {
   // Doing some asynchronous checking then set safe mode
}

const processData = ({ setSafeMode }) => {
   setSafeMode();
   doSomethingWithSafeMode();
}

```

=> The potential bug might happen because doSomethingWithSafeMode() is triggered before safe mode is being evaluated.

That's not the case when using with synthetic event like `onSubmit`, `onClick`, `onChange`, `onBlur` ...

If you see any `leaking a Promise (i.e. errors produced by the promise are not handled)` that can cause wrong logic handling by using async await on those event handlers, like:

```
<form onSubmit={async () => {
    await doSomething();
}}>
  ...
</form>

```

Please share it here to discuss. Otherwise, just use the config from https://github.com/typescript-eslint/typescript-eslint/pull/4623 or wrapper function is enough.

The config example:
```
{
  "@typescript-eslint/no-misused-promises": [
    "error",
    {
      "checksVoidReturn": {
        "arguments": false
      }
    }
  ]
}
```

**Reply by kdubb** [reply]
Before we go further into that, can you explain the reasoning for `handleSubmit` generating a `Promise` returning function when `onSubmit` is typed to accept a `void` returning function? This seems non-sensical and of no added value.

By "fixing" `handleSubmit` to return a `void` returning function it will force anybody with this rule enabled to explicitly handle the `Promise`, which is what they are asking to be linted. For codebases without this rule enabled it seems nothing changes.

Asking to disable/diminish a pretty useful rule because `handleSubmit` is producing a function with the wrong signature seems odd.

**Reply by leapful** [reply]
@kdubb 

Method `handleSubmit` is using `async await` to perform async validation logic and also trigger `onValid` or `onError` callback which can be an async callback too. 

https://github.com/react-hook-form/react-hook-form/blob/1647bb2295ef4f7633b0e30c2158dac9e0a2898a/src/logic/createFormControl.ts#L1027-L1077

Since there is nothing wrong with it, changing might not bed needed. Furthermore, using `async await` on DOM event handler is a common pattern in React and updating a linter configuration is easier to solve this specific case.

**Reply by kdubb** [reply]
I must be missing something. Shouldn't `handleSubmit` catch those errors? Adding a `catch` clause to the currently returned function would ensure those errors update the form state accordingly. Then `handleSubmit` can return a `void` returning function wrapping that logic.

**Reply by paleite** [reply]
I've created [a PR for this](https://github.com/react-hook-form/react-hook-form/pull/10924), which shows exactly where the issue is problematic. In the tests, it's assumed that if onValid throws an error, you need to catch it after handleSubmit has already been executed, i.e. if you use handleSubmit like the docs suggest, you're leaking a promise.

Case in point: https://github.com/react-hook-form/react-hook-form/pull/10924/files#annotation_14103380485

**Comment by demarchenac** [comment]
Are there any updates on the go-to implementation one should follow?

**Reply by demarchenac** [reply]
I'm asking since there seems to be an accepted answer on a similar [discussion #9325](https://github.com/react-hook-form/react-hook-form/discussions/9325#discussioncomment-4060566):

> Alright, I've done a little bit more research.
> 
> Check out this [typescript-eslint/typescript-eslint#4623](https://github.com/typescript-eslint/typescript-eslint/pull/4623) PR that adds granular options within `checksVoidReturns` for [the `no-misused-promises` TS ESLint rule](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/no-misused-promises.md); the `attributes` options was added specifically for use-cases like the one you're experiencing. This is the official TS ESLint solution and we recommend you update your `.eslintrc` config file with it:
> 
> ```json
> "rules": {
>   "@typescript-eslint/no-misused-promises": [2, {
>     "checksVoidReturn": {
>       "attributes": false
>     }
>   }]
> },
> ```

**Comment by xenostar** [comment]
We need to continue the discussion around this. I just installed `react-hook-form` tonight and within 5 minutes of using it I'm jumping down a huge rabbit hole trying to "fix" this issue. I imagine many others are in the same situation considering this library and at this moment it almost feels more convenient just to uninstall it. That is a shame.

Changing my `eslint` rules is in my opinion an absurd "solution" to this problem since it is flagging a legitimate issue; this needs to be solved at the `react-hook-form` level.

I suggest we either:

1. Change the `handleSubmit` function implementation to be wrapped in one of the many suggested promise helper functions outlined here (and in #9325 and #8622) that handles the promise correctly.
2. Or, provide this wrapper function so it can be imported and we can wrap our `handleSubmit` with it. i.e.
```tsx
<form onSubmit={handleSubmitHelperFunction(handleSubmit(onSubmit)} />
```

Personally, I vote for option 1 because I think users shouldn't have to think about this issue.

**Reply by ahriannn** [reply]
I agree with solution 1, it will make it easier for people to update their codebase if for example the eslint rule was added at a later date etc. Doing a previous comments hack worked best for me:
`(...args) => void handleSubmit(onSubmit)(...args)`

**Reply by bombillazo** [reply]
Has this issue been solved?

**Reply by gheneca** [reply]
No?

**Reply by SalahAdDin** [reply]
Let's create an issue then!

**Comment by bostrom** [comment]
FWIW I was using @paleite 's [workaround](https://github.com/orgs/react-hook-form/discussions/8020#discussioncomment-3362300) for a while until noticing that the `formState` sometimes got reset when submitting the form multiple times (e.g. due to validation errors or error responses from the backend.)

I started using this instead, which has been working better


```ts
  import { FormEventHandler, useCallback } from 'react';
  import { FieldValues, useForm } from 'react-hook-form'; 
  import { Form, useSubmit } from 'react-router-dom';

  const submit = useSubmit(); // from react-router
  const { handleSubmit } =  useForm<SomeFormType>({
    // ...
  })

  const onSubmit = useCallback<FormEventHandler>(
    (event) =>
      handleSubmit((values: FieldValues) => {
        submit(values, {
          // submit your values somehow, here using react-router's submit
        });
      })(event),
    [handleSubmit, submit],
  );

  return (
    <Form
      method="post"
      onSubmit={onSubmit}
    >
      // ....
    </Form>
  );

```
